tidy3d/components/monitor.py

```diff
@@ -447,6 +383,247 @@ def storage_size(self, num_cells: int, tmesh: int) -> int:
+                "valid for box monitors only."
+            )
+        return values
+
+    @property
+    def axis(self) -> Axis:
+        """Returns the :class:`.Axis` normal to this surface."""
+        # assume that the monitor's axis is in the direction where the monitor is thinnest
+        # and errors if the monitor is a box
+        if self.size.count(0.0) != 1:
+            raise DataError(
+                "Requested ``axis`` property for a box monitor; ``axis`` is defined "
+                "for surface monitors only."
+            )
+        return self.size.index(0.0)
+
+    @property
+    def local_origin(self) -> Coordinate:
+        """Returns the local origin associated with this monitor."""
+        if self.custom_origin is None:
+            return self.center
+        return self.custom_origin
+
+
+class Near2FarAngleMonitor(AbstractNear2FarMonitor):
+    """:class:`Monitor` that samples electromagnetic near fields in the frequency domain
+       and invokes the computation of far fields at given observation angles.
+
+    Example
+    -------
+    >>> monitor = Near2FarAngleMonitor(
+    ...     center=(1,2,3),
+    ...     size=(2,2,2),
+    ...     freqs=[250e12, 300e12],
+    ...     name='n2f_monitor',
+    ...     custom_origin=(1,2,3),
+    ...     phi=[0, np.pi/2],
+    ...     theta=np.linspace(-np.pi/2, np.pi/2, 100)
+    ...     )
+    """
+
+    _data_type: Literal["Near2FarData"] = pydantic.Field("Near2FarAngleData")
+
+    theta: ObsGridArray = pydantic.Field(
+        ...,
+        title="Polar Angles",
+        description="Polar angles relative to ``local_origin`` at which to compute far fields.",
+        units=RADIAN,
+    )
+
+    phi: ObsGridArray = pydantic.Field(
+        ...,
+        title="Azimuth Angles",
+        description="Azimuth angles relative to ``local_origin`` at which to compute far fields.",
+        units=RADIAN,
+    )
+
+    def storage_size(self, num_cells: int, tmesh: ArrayLike[float, 1]) -> int:
+        # stores 1 complex number per pair of angles, per frequency,
+        # for N_theta, N_phi, L_theta, and L_phi (4 components)
+        return BYTES_COMPLEX * len(self.theta) * len(self.phi) * len(self.freqs) * 4
+
+
+class Near2FarCartesianMonitor(AbstractNear2FarMonitor):
+    """:class:`Monitor` that samples electromagnetic near fields in the frequency domain
+       and invokes the computation of far fields on a Cartesian observation plane.
+
+    Example
+    -------
+    >>> monitor = Near2FarCartesianMonitor(
+    ...     center=(1,2,3),
+    ...     size=(2,2,2),
+    ...     freqs=[250e12, 300e12],
+    ...     name='n2f_monitor',
+    ...     custom_origin=(1,2,3),
+    ...     x=[-1, 0, 1],
+    ...     y=[-2, -1, 0, 1, 2],
+    ...     plane_axis=2,
+    ...     plane_distance=5
+    ...     )
+    """
+
+    _data_type: Literal["Near2FarCartesianData"] = pydantic.Field("Near2FarCartesianData")
```

Comment:
we dont need these as long as the monitor -> data mapping is in `DATA_TYPE_MAP`, can remove